Add Qdrant container

russcam opened the pull request:
Closes #992

## What does this PR do?

This commit adds a qdrant container to the list of supported Testcontainers.

The qdrant container allows configuration of:

- an API key to authenticate to Qdrant
- an x509 certificate used to secure communication to Qdrant with Transport Layer Security
- a custom configuration file. See https://qdrant.tech/documentation/guides/configuration/

## Why is it important?

Provides a qdrant test container for folks that need it

## Related issues

#992

## How to test this PR

Run tests in Testcontainers.Qdrant.Tests

russcam committed:
Add Qdrant container

This commit adds a qdrant container to the list of supported
Testcontainers.

The qdrant container allows configuration of:

- an API key to authenticate to Qdrant
- an x509 certificate used to secure communication to Qdrant with
  Transport Layer Security
- a custom configuration file. See
  https://qdrant.tech/documentation/guides/configuration/

Closes #992

russcam committed:
Fix TLS tests on Linux

Qdrant uses rustls for TLS, which does not accept IPv4 or IPv6 addresses as the hostname in SNI, adhering
to RFC 6066: https://www.rfc-editor.org/rfc/rfc6066#page-7

Sending an IPv4 or IPv6 address as the hostname in SNI causes the handshake to be rejected and
qdrant to log

WARN rustls::msgs::handshake: Illegal SNI hostname received "127.0.0.1"

By default, .NET on linux sends the DNS name in the URI in Server Name Indication (SNI),
irrespective of whether it's an IP address or not. In order to resolve this, a custom
Host request header is added, per

https://github.com/dotnet/runtime/issues/20876#issuecomment-379585707

russcam committed:
Remove superfluous solution items

russcam commented:
This is ready to review @HofmeisterAn

HofmeisterAn commented on src/Testcontainers.Qdrant/QdrantBuilder.cs:
```diff
@@ -0,0 +1,109 @@
+using System.IO;
+using System.Security.Cryptography.X509Certificates;
+using Org.BouncyCastle.OpenSsl;
+using Org.BouncyCastle.Security;
```
Please move the usings (other classes too) into the global `Usings.cs` file.

HofmeisterAn commented on src/Testcontainers.Qdrant/QdrantBuilder.cs:
```diff
@@ -0,0 +1,109 @@
+	/// </summary>
+	/// <param name="configurationFilePath">The path to the configuration file</param>
+	public QdrantBuilder WithConfigFile(string configurationFilePath) =>
+		Merge(DockerResourceConfiguration, new QdrantConfiguration(configurationFilePath: configurationFilePath))
```
If we do not require the `ConfigurationFilePath` property, along with the other configuration properties, at some time later, such as within the Qdrant container instance, then there is no need to keep and store the value. I think we can remove those properties from `QdrantConfiguration`

HofmeisterAn commented on src/Testcontainers.Qdrant/QdrantBuilder.cs:
```diff
@@ -0,0 +1,109 @@
+	/// <param name="configurationFilePath">The path to the configuration file</param>
+	public QdrantBuilder WithConfigFile(string configurationFilePath) =>
+		Merge(DockerResourceConfiguration, new QdrantConfiguration(configurationFilePath: configurationFilePath))
+			.WithBindMount(configurationFilePath, QdrantLocalConfigurationFilePath);
```
```suggestion
			.WithResourceMapping(configurationFilePath, QdrantLocalConfigurationFilePath);
```

HofmeisterAn commented on src/Testcontainers.Qdrant/QdrantBuilder.cs:
```diff
@@ -0,0 +1,109 @@
+	/// private key.
+	/// </summary>
+	/// <param name="certificate">A certificate containing a private key</param>
+	public QdrantBuilder WithCertificate(X509Certificate2 certificate)
```
Can we not expect the developer to provide valid `cert.pem` and `key.pem` files? This way, we can avoid dealing with all the certificate-related issues and Bouncy Castle.

HofmeisterAn commented on tests/Testcontainers.Qdrant.Tests/Testcontainers.Qdrant.Tests.csproj:
```diff
@@ -0,0 +1,17 @@
+        <IsPublishable>false</IsPublishable>
+    </PropertyGroup>
+    <ItemGroup>
+        <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.4.1"/>
+        <PackageReference Include="coverlet.collector" Version="3.2.0"/>
+        <PackageReference Include="xunit.runner.visualstudio" Version="2.4.5"/>
+        <PackageReference Include="xunit" Version="2.4.2"/>
```
```suggestion
        <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.7.2"/>
        <PackageReference Include="coverlet.collector" Version="6.0.0"/>
        <PackageReference Include="xunit.runner.visualstudio" Version="2.5.0"/>
        <PackageReference Include="xunit" Version="2.5.0"/>
```

HofmeisterAn reviewed (commented):
> This is ready to review @HofmeisterAn

Thank you for your pull request. I've already seen the PR but haven't had the time to thoroughly review it yet. Spare time is limited 😅.

Most of it looks good! I just have a few minor suggestions and improvements.

In general, we don't expose an API for every configuration (e.g., environment variable) if the developer can configure it themselves through the generic API (it doesn't require in-depth knowledge or isn't crucial for starting a basic container). However, I'm okay with the exposed APIs, but I would like to clean up some parts related to the TLS implementation if needed. I'll need to take a closer look at it (I will look into it by end of this week).

HofmeisterAn commented:
This pull request has been inactive for some time. I will close it for now. If you wish to continue working on it, please feel free to reopen it.

russcam commented:
@HofmeisterAn please can we reopen this PR; I've rebased from `develop` and updated following PR feedback. If it's easier, I can always open a new PR, but I don't have the ability to reopen this one

HofmeisterAn commented:
> @HofmeisterAn please can we reopen this PR; I've rebased from `develop` and updated following PR feedback. If it's easier, I can always open a new PR, but I don't have the ability to reopen this one

I am sorry, I cannot reopen it either, it says:

> The qdrant-container branch was force-pushed or recreated.

You probably need to create a new PR - sorry.

russcam commented:
Superseded by #1149
